lib/matplotlib/colorizer.py

```diff
@@ -0,0 +1,483 @@
+"""
+The Colorizer class which handles the data to color pipeline via a
+normalization and a colormap.
+
+.. seealso::
+
+  :doc:`/gallery/color/colormap_reference` for a list of builtin colormaps.
+
+  :ref:`colormap-manipulation` for examples of how to make
+  colormaps.
+
+  :ref:`colormaps` an in-depth discussion of choosing
+  colormaps.
+
+  :ref:`colormapnorms` for more details about data normalization.
+"""
+
+import numpy as np
+from numpy import ma
+import functools
+from matplotlib import _api, colors, cbook, scale, cm
+
+
+class Colorizer():
+    """
+    Class that holds the data to color pipeline
+    accessible via `.to_rgba(A)` and executed via
+    the `.norm` and `.cmap` attributes.
+    """
+    def __init__(self, cmap=None, norm=None):
+
+        self._cmap = None
+        self._set_cmap(cmap)
+
+        self._id_norm = None
+        self._norm = None
+        self.norm = norm
+
+        self.callbacks = cbook.CallbackRegistry(signals=["changed"])
+        self.colorbar = None
+
+    def _scale_norm(self, norm, vmin, vmax, A):
+        """
+        Helper for initial scaling.
+
+        Used by public functions that create a ScalarMappable and support
+        parameters *vmin*, *vmax* and *norm*. This makes sure that a *norm*
+        will take precedence over *vmin*, *vmax*.
+
+        Note that this method does not set the norm.
+        """
+        if vmin is not None or vmax is not None:
+            self.set_clim(vmin, vmax)
+            if isinstance(norm, colors.Normalize):
+                raise ValueError(
+                    "Passing a Normalize instance simultaneously with "
+                    "vmin/vmax is not supported.  Please pass vmin/vmax "
+                    "directly to the norm when creating it.")
+
+        # always resolve the autoscaling so we have concrete limits
+        # rather than deferring to draw time.
+        self.autoscale_None(A)
+
+    @property
+    def norm(self):
+        return self._norm
+
+    @norm.setter
+    def norm(self, norm):
+        _api.check_isinstance((colors.Normalize, str, None), norm=norm)
+        if norm is None:
+            norm = colors.Normalize()
+        elif isinstance(norm, str):
+            try:
+                scale_cls = scale._scale_mapping[norm]
+            except KeyError:
+                raise ValueError(
+                    "Invalid norm str name; the following values are "
+                    f"supported: {', '.join(scale._scale_mapping)}"
+                ) from None
```

Comment:
Rather https://matplotlib.org/devdocs/api/_api_api.html#matplotlib._api.check_getitem

Reply:
This code mirrors the existing code for ScalarMappable (cm.py lines 545→552), but I'll see if I can shift it to use the _api or if there is some reason why that has not been implemented for ScalarMappable.

Reply:
Two points:
1. This will change the error message from:
```
" Invalid norm str name; the following values are supported: 'linear', 'log', 'symlog', 'asinh', 'logit', 'function', 'functionlog'
``` 
to
```
' ' is not a valid value for norm; supported values are 'linear', 'log', 'symlog', 'asinh', 'logit', 'function', 'functionlog'
``` 
Which is slightly misleading, as norm supports other inputs than strings. We can mitigate this by intercepting the error message and replacing **value** with **string**, but this requires an additional try except block.

2. I think this check naturally belongs in `_auto_norm_from_scale`, a function which is currently called by:
```py
        elif isinstance(norm, str):
            try:
                scale_cls = scale._scale_mapping[norm]
            except KeyError:
                raise ValueError(
                    "Invalid norm str name; the following values are "
                    f"supported: {', '.join(scale._scale_mapping)}"
                ) from None
            norm = _auto_norm_from_scale(scale_cls)()
```
which I think should be replaced by:
```py
        elif isinstance(norm, str):
            norm = _auto_norm_from_str(norm)()
```
(note the change in function name from `_scale` to `_str`)
This function then incorporates the check.
```py
def _auto_norm_from_str(norm_str):
    scale_cls = _api.check_getitem(scale._scale_mapping, norm=norm_str)
    ...
```
@timhoffm @story645 Let me know your opinions on point 1 and 2 above, and I will implement it in this PR.
Point 2. would be a useful change with regards to making a 'MultiNorm' in a later PR.

Reply:
I agree the message would be slightly misleading. OTOH pushing this down does not make much better for the user. Thy still get the same exception message. Only the call stack is different. While the message in the deeper context is correct, it's further away from the public API and thus more difficult to trace back.

An alternative would be expanding `_check_get_item` so that we can do `_api.check_getitem(scale._scale_mapping, norm=norm_str, _type="str")` and get "' ' is not a valid str value ...", which would imply there may be other valid types.

I believe in practice, it does not matter. If the user has passed an invalid str, they most likely have a typo and still wanted to pass a str name. It's unlikely that the information being able to pass Norms directly would be helpful. To keep things simple, I slighly favor just using `_check_get_item` as is in original context. But no strong opinion.

Reply:
I think it's better to have it here, with the changes in #28690 that means we only have to move this function once.